upstream/registry.go

```diff
@@ -153,16 +153,40 @@ func (u *UpstreamsRegistry) PrepareUpstreamsForNetwork(ctx context.Context, netw
 		close(errCh)
 	}()
 
-	// Wait for at least one upstream, completion, or error
+	// Wait for at least one active upstream, completion, or error with a 3-second timeout
 	ticker := time.NewTicker(1 * time.Second)
 	defer ticker.Stop()
 
+	timeout := time.NewTimer(3 * time.Second)
+	defer timeout.Stop()
+
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		case err := <-errCh:
 			return err
+		case <-timeout.C:
+			// Timeout expired - check if we have any active upstreams
+			u.upstreamsMu.RLock()
+			upstreams := u.networkUpstreams[networkId]
+			activeCount := 0
+			for _, ups := range upstreams {
+				if !u.metricsTracker.IsCordoned(ups, "*") {
+					activeCount++
+				}
+			}
```

Comment:
There should better ways to assert being active/healthy. The cordoning is only if selection policy is active and it had a chance to do a evaluation at least once.

The fact of upstream being in "u.networkUpstreams" means it's already initialized. But how many are necessary to return successfully is debatable.

---

P.S. The ticket.C section is already taking care of asserting active upstreams and <-ctx.Done(): already takes care of timeout (i.e. if it takes long), we don't need this <-timeout.C case afaict.